nixos/acme: harden systemd units

mweinelt opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

This change applies some hopefully non-controversial hardening. Hardening is always a fickle business and I hope I didn't step on anyones usecase.

I've run through the acme test suite and it is still working :grinning: 

```
✗ PrivateNetwork=                                             Service has access to the host's network                                                                 0.5
✗ RestrictAddressFamilies=~AF_(INET|INET6)                    Service may allocate Internet sockets                                                                    0.3
✗ DeviceAllow=                                                Service has a device ACL with some special devices                                                       0.1
✗ IPAddressDeny=                                              Service does not define an IP address allow list                                                         0.2
✗ PrivateUsers=                                               Service has access to other users                                                                        0.2
✗ ProtectSystem=                                              Service has very limited write access to the OS file hierarchy                                           0.1
✗ SystemCallFilter=~@privileged                               System call allow list defined for service, and @privileged is included (e.g. chown is allowed)          0.2
✗ SystemCallFilter=~@resources                                System call allow list defined for service, and @resources is included (e.g. ioprio_set is allowed)      0.2
✗ RootDirectory=/RootImage=                                   Service runs within the host's root directory                                                            0.1
✗ UMask=                                                      Files created by service are world-readable by default                                                   0.1

→ Overall exposure level for acme-a.example.test.service: 1.5 OK :-)
```

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

mweinelt committed:
nixos/acme: harden systemd units

mweinelt commented on nixos/modules/security/acme.nix:
```diff
@@ -21,15 +21,41 @@ let
+    RestrictRealtime = true;
+    RestrictSUIDSGID = true;
+    SystemCallArchitectures = "native";
+    SystemCallFilter = "@system-service ~@resources ~@ipc ~@keyring ~@resources ~@setuid";
```
This is the most controversial part honestly.

As I understand it, in a mixed allow/deny list the evaluation goes from left to right and adds and removes things in order. So I add `@system-service`, then remove `@resources`, `@ipc`, etc. from it.

> If you specify both types of this option (i.e. allow-listing and deny-listing), the first encountered will take precedence and will dictate the default action (termination or approval of a system call). Then the next occurrences of this option will add or delete the listed system calls from the set of the filtered system calls, depending of its type and the default action. (For example, if you have started with an allow list rule for read() and write(), and right after it add a deny list rule for write(), then write() will be removed from the set.)

Honestly I'm now not sure, why `systemd-analyze security` still shows I allow `@privileged`. I'm not worried about `@privileged`, because we actually allow `@chown`.

m1cr0man replied:
Is it intentional to have `~@resources` there twice? Also how does this affect a call to systemctl (for example to restart another service as configured in `security.acme.certs.<cert>.postRun`)?

mweinelt replied:
> Is it intentional to have ~@resources there twice?

Indeed unintentional, thanks for spotting this.

> Also how does this affect a call to systemctl (for example to restart another service as configured in security.acme.certs.<cert>.postRun)?

This is what is in `@resources`, so I think not at all.

```
❯ systemd-analyze syscall-filter @resources
@resources
    # Alter resource settings
    ioprio_set
    mbind
    migrate_pages
    move_pages
    nice
    sched_setaffinity
    sched_setattr
    sched_setparam
    sched_setscheduler
    set_mempolicy
    setpriority
    setrlimit
```

Unless you meant the syscall filter in general. I think `@process`, which is part of `@system-service` takes care of that:

```
❯ systemd-analyze syscall-filter @process
@process
    # Process control, execution, namespacing operations
    arch_prctl
    capget
    clone
    clone3
    execveat
    fork
    getrusage
    kill
    pidfd_open
    pidfd_send_signal
    prctl
    rt_sigqueueinfo
    rt_tgsigqueueinfo
    setns
    swapcontext
    tgkill
    times
    tkill
    unshare
    vfork
    wait4
    waitid
    waitpid
``

mweinelt replied:
Remove the duplicate `@resources` deny entry.

mweinelt commented on nixos/modules/security/acme.nix:
```diff
@@ -21,15 +21,41 @@ let
+    Type = "oneshot";
+    User = "acme";
+    Group = mkDefault "acme";
+    UMask = 0022;
```
```
✗ UMask=                                                      Files created by service are world-readable by default                                                   0.1
```

I'm not sure, but can't be default to `0027` (640, 750) instead?

m1cr0man replied:
Yeah I've done that in https://github.com/NixOS/nixpkgs/pull/121750/files.

mweinelt replied:
Sadly, this interferes with access to http-01 challenges

m1cr0man replied:
Ah also #106603 which I almost forgot about!

m1cr0man commented:
I'm getting a warning on my system after applying these changes:
```
May 16 18:27:37 myserv systemd[1]: acme-example.com.service:48: Failed to parse system call, ignoring: ~@ipc
May 16 18:27:37 myserv systemd[1]: acme-example.com.service:48: Failed to parse system call, ignoring: ~@keyring
May 16 18:27:37 myserv systemd[1]: acme-example.com.service:48: Failed to parse system call, ignoring: ~@resources
May 16 18:27:37 myserv systemd[1]: acme-example.com.service:48: Failed to parse system call, ignoring: ~@setuid
```
Is the syntax correct? I don't know what it should be personally.

mweinelt commented:
Hm, so it seems that is not the way to remove individual syscall groups after all. I'll be checking up the docs … again.

mweinelt commented:
@m1cr0man I've updated the `SystemCallFilter` and added some comments.

I also added `ProtectSystem = "strict";`, so now paths you want to access need to be added to `ReadOnlyPaths` or `ReadWritePaths`. Does that sound reasonable to you?

```
✗ PrivateNetwork=                                             Service has access to the host's network                                                             0.5
✗ RestrictAddressFamilies=~AF_(INET|INET6)                    Service may allocate Internet sockets                                                                0.3
✗ DeviceAllow=                                                Service has a device ACL with some special devices                                                   0.1
✗ IPAddressDeny=                                              Service does not define an IP address allow list                                                     0.2
✗ PrivateUsers=                                               Service has access to other users                                                                    0.2
✗ SystemCallFilter=~@privileged                               System call allow list defined for service, and @privileged is included (e.g. chown is allowed)      0.2
✗ RootDirectory=/RootImage=                                   Service runs within the host's root directory                                                        0.1
✗ UMask=                                                      Files created by service are world-readable by default                                               0.1

→ Overall exposure level for acme-a.example.test.service: 1.4 OK :-)
```

m1cr0man reviewed (approved):
> @m1cr0man I've updated the `SystemCallFilter` and added some comments.
> 
> I also added `ProtectSystem = "strict";`, so now paths you want to access need to be added to `ReadOnlyPaths` or `ReadWritePaths`. Does that sound reasonable to you?

Yeah so I've just done some testing with this. We [prefix the postRun](https://github.com/NixOS/nixpkgs/blob/ca0f0212c2623822d7fa0254e844a1f787498f01/nixos/modules/security/acme.nix#L264) scripting with "+" which [from the docs](https://www.freedesktop.org/software/systemd/man/systemd.service.html#ExecStart=) means it ignores most of the hardening being imposed here. 

I modified the postRun test such that it writes to a file outside of /var/lib/acme and it was successful. This is something I would be concerned about breaking in existing setups but with this test I would be confident that they won't be affected.

Happy to see it merged now, but if you could apply this patch so that the test suite covers what I'm explaining above that would be great.
```diff
diff --git a/nixos/tests/acme.nix b/nixos/tests/acme.nix
index 6f98b0da378..edd387cae0c 100644
--- a/nixos/tests/acme.nix
+++ b/nixos/tests/acme.nix
@@ -105,9 +105,9 @@ in import ./make-test-python.nix ({ lib, ... }: {
         security.acme.certs."a.example.test".keyType = "ec384";
         security.acme.certs."a.example.test".postRun = ''
           set -euo pipefail
-          touch test
-          chown root:root test
-          echo testing > test
+          touch /home/test
+          chown root:root /home/test
+          echo testing > /home/test
         '';
       };
 
@@ -375,7 +375,7 @@ in import ./make-test-python.nix ({ lib, ... }: {
           switch_to(webserver, "cert-change")
           webserver.wait_for_unit("acme-finished-a.example.test.target")
           check_connection_key_bits(client, "a.example.test", "384")
-          webserver.succeed("grep testing /var/lib/acme/a.example.test/test")
+          webserver.succeed("grep testing /home/test")
 
       with subtest("Correctly implements OCSP stapling"):
           switch_to(webserver, "ocsp-stapling")
```

m1cr0man committed:
nixos/tests/acme: test access to files outside /var/lib/acme in postRun

mweinelt commented:
Rebased after merging #121750.
